src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2153,59 +2153,57 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
             else if (Vector256.IsHardwareAccelerated && length >= Vector256<TValue>.Count)
             {
                 Vector256<TValue> equals, values = Vector256.Create(value);
-                ref TValue currentSearchSpace = ref Unsafe.Add(ref searchSpace, length - Vector256<TValue>.Count);
+                nint offset = length - Vector256<TValue>.Count;
 
-                // Loop until either we've finished all elements or there's less than a vector's-worth remaining.
-                do
+                // Loop until either we've finished all elements -or- there's one or less than a vector's-worth remaining.
+                while (offset > 0)
                 {
-                    equals = TNegator.NegateIfNeeded(Vector256.Equals(values, Vector256.LoadUnsafe(ref currentSearchSpace)));
+                    equals = TNegator.NegateIfNeeded(Vector256.Equals(values, Vector256.LoadUnsafe(ref searchSpace, (nuint)(offset))));
+
                     if (equals == Vector256<TValue>.Zero)
                     {
-                        currentSearchSpace = ref Unsafe.Subtract(ref currentSearchSpace, Vector256<TValue>.Count);
+                        offset -= Vector256<TValue>.Count;
                         continue;
                     }
 
-                    return ComputeLastIndex(ref searchSpace, ref currentSearchSpace, equals);
+                    return ComputeLastIndex(offset, equals);
                 }
-                while (!Unsafe.IsAddressLessThan(ref currentSearchSpace, ref searchSpace));
 
                 // If any elements remain, process the first vector in the search space.
```

Comment:
Nit: this is now processing the first vector in the search space regardless of whether any elements remain.  Consider tweaking the comment to remove the condition.